File: front/src/types/router.ts
/**
 * Tipos para el sistema de gestión de routers MikroTik
 */

export interface Branch {
  id: number;
  name: string;
  location?: string;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface Router {
  id: number;
  name: string;
  uri: string;
  username: string;
  password: string;
  branchId: number;
  branch?: Branch;
  isActive: boolean;
  status?: 'online' | 'offline' | 'error';
  createdAt: string;
  updatedAt: string;
}

export interface PPPoEClient {
  id: string;
  routerId: number;
  router?: Router;
  ip?: string;
  ipAddress?: string;
  name: string;
  password: string;
  comment?: string;
  profile?: string;
  profileValue?: string;
  contract?: string;
  isActive: boolean;
  status?: 'active' | 'suspended' | 'blocked' | 'disconnected';
  createdAt: string;
  updatedAt?: string;
}

export interface PPPoESession {
  id: string;
  clientId: string;
  client?: PPPoEClient;
  ip?: string;
  ipAddress?: string;
  uptime: string;
  bytesIn: number;
  bytesOut: number;
  packetsIn: number;
  packetsOut: number;
  createdAt: string;
}

export interface RouterInterface {
  id: string;
  routerId: number;
  name: string;
  type: string;
  status: 'up' | 'down';
  macAddress?: string;
  mtu?: number;
  rxBytes: number;
  txBytes: number;
  rxPackets: number;
  txPackets: number;
}

export interface RouterResource {
  routerId: number;
  cpuLoad: number;
  memoryUsed: number;
  memoryTotal: number;
  diskUsed: number;
  diskTotal: number;
  uptime: string;
  temperature?: number;
  voltage?: number;
  timestamp: string;
}

export interface FirewallRule {
  id: string;
  routerId: number;
  router?: Router;
  firewallId: string;
  ipAddress: string;
  comment?: string;
  creationDate?: string;
  isActive: boolean;
  protocol?: string;
  port?: string;
  action?: string;
  chain?: string;
}

export interface DashboardStats {
  totalRouters: number;
  onlineRouters: number;
  totalClients: number;
  activeClients: number;
  suspendedClients: number;
  blockedClients: number;
  totalSessions: number;
  totalBandwidth: number;
}

export interface Alert {
  id: string;
  type: 'error' | 'warning' | 'info';
  title: string;
  message: string;
  routerId?: number;
  router?: Router;
  clientId?: string;
  client?: PPPoEClient;
  isRead: boolean;
  createdAt: string;
}

export interface ActivityLog {
  id: number;
  routerId: number;
  router?: Router;
  userId: number;
  user?: unknown;
  action: string;
  description?: string;
  ipAddress?: string;
  userAgent?: string;
  createdAt: string;
}

export interface SyncLog {
  id: string;
  routerId: number;
  router?: Router;
  action: 'sync' | 'update' | 'delete';
  status: 'success' | 'error';
  message?: string;
  itemsProcessed: number;
  createdAt: string;
}

// Request/Response types
export interface CreateRouterRequest {
  name: string;
  uri: string;
  username: string;
  password: string;
  branchId: number;
}

export interface UpdateRouterRequest {
  name?: string;
  uri?: string;
  username?: string;
  password?: string;
  branchId?: number;
  isActive?: boolean;
}

export interface CreatePPPoEClientRequest {
  routerId: number;
  name: string;
  password: string;
  comment?: string;
  profile?: string;
  contract?: string;
}

export interface UpdatePPPoEClientRequest {
  name?: string;
  password?: string;
  comment?: string;
  profile?: string;
  contract?: string;
  isActive?: boolean;
}

export interface ClientStatusUpdate {
  status: 'active' | 'suspended' | 'blocked';
  reason?: string;
}

export interface BlockIPRequest {
  ipAddress: string;
  comment?: string;
}

export interface ClientBandwidthStats {
  clientId: string;
  bytesIn: number;
  bytesOut: number;
  packetsIn: number;
  packetsOut: number;
  timestamp: string;
}

export interface ClientStats {
  clientId: string;
  totalSessions: number;
  totalUptime: string;
  avgBandwidthIn: number;
  avgBandwidthOut: number;
  lastConnection: string;
  totalBytes: number;
}

// API Response types
export interface ApiResponse<T> {
  success: boolean;
  data?: T;
  message?: string;
  error?: string;
}

export interface PaginatedResponse<T> {
  success: boolean;
  data: T[];
  pagination: {
    page: number;
    limit: number;
    total: number;
    totalPages: number;
  };
}

// Search and Filter types
export interface RouterFilters {
  search?: string;
  branchId?: number;
  isActive?: boolean;
  status?: 'online' | 'offline' | 'error';
}

export interface PPPoEClientFilters {
  search?: string;
  routerId?: number;
  status?: 'active' | 'suspended' | 'blocked' | 'disconnected';
  profile?: string;
  isActive?: boolean;
}

export interface SessionFilters {
  routerId?: number;
  clientId?: string;
  ip?: string;
  ipAddress?: string;
}

File: back/models/router.py
from datetime import datetime
from models import db
from .base import BaseModel


class Branch(BaseModel):
    """Modelo Branch basado en tabla 'branches'"""
    __tablename__ = 'branches'
    
    id = db.Column(db.Integer, primary_key=True)
    name = db.Column(db.String(255), nullable=False)
    location = db.Column(db.String(255), nullable=True)
    is_active = db.Column(db.Boolean, default=True)
    
    # Relaciones
    routers = db.relationship('Router', back_populates='branch', lazy='dynamic')


class Router(BaseModel):
    """Modelo Router basado en tabla 'routers'"""
    __tablename__ = 'routers'
    
    id = db.Column(db.Integer, primary_key=True)
    name = db.Column(db.String(100), nullable=False)
    uri = db.Column(db.String(255), nullable=False)
    username = db.Column(db.String(100), nullable=False)
    password = db.Column(db.String(255), nullable=False)
    branch_id = db.Column(db.Integer, db.ForeignKey('branches.id'), nullable=False)
    is_active = db.Column(db.Boolean, default=True)
    
    # Relaciones
    branch = db.relationship('Branch', back_populates='routers')
    user_routers = db.relationship('UserRouter', back_populates='router', lazy='dynamic')
    secrets = db.relationship('Secret', back_populates='router', lazy='dynamic')
    firewall_rules = db.relationship('RouterFirewall', back_populates='router', lazy='dynamic')
    secret_configs = db.relationship('RouterSecretConfig', back_populates='router', lazy='dynamic')
    activity_logs = db.relationship('ActivityLog', back_populates='router', lazy='dynamic')


class Secret(db.Model):
    """Modelo RouterSecret basado en tabla 'router_secrets'"""
    __tablename__ = 'router_secrets'
    
    id = db.Column(db.Integer, primary_key=True)
    router_id = db.Column(db.Integer, db.ForeignKey('routers.id'), nullable=False)
    ip_address = db.Column(db.String(20), nullable=False)
    name = db.Column(db.String(100), nullable=False)
    password = db.Column(db.String(255), nullable=False)
    comment = db.Column(db.String(255), nullable=True)
    profile = db.Column(db.String(255), nullable=True)
    contract = db.Column(db.String(255), nullable=True)
    is_active = db.Column(db.Boolean, default=True)
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    
    # Relaciones
    router = db.relationship('Router', back_populates='secrets')


class RouterFirewall(db.Model):
    """Modelo RouterFirewall basado en tabla 'router_firewall'"""
    __tablename__ = 'router_firewall'

    router_id = db.Column(db.Integer, db.ForeignKey('routers.id'), primary_key=True)
    firewall_id = db.Column(db.String(50), primary_key=True)
    ip_address = db.Column(db.String(20), nullable=False)
    comment = db.Column(db.String(255), nullable=True)
    creation_date = db.Column(db.String(100), nullable=True)
    protocol = db.Column(db.String(20), nullable=True)
    port = db.Column(db.String(255), nullable=True)
    action = db.Column(db.String(20), nullable=True)
    chain = db.Column(db.String(20), nullable=True)
    is_active = db.Column(db.Boolean, default=True)
    created_at = db.Column(db.DateTime, default=datetime.utcnow)

    # Relaciones
    router = db.relationship('Router', back_populates='firewall_rules')


class RouterSecretConfig(db.Model):
    """Modelo RouterSecretConfig basado en tabla 'router_secret_configs'"""
    __tablename__ = 'router_secret_configs'
    
    router_id = db.Column(db.Integer, db.ForeignKey('routers.id'), primary_key=True)
    secret_id = db.Column(db.String(50), primary_key=True)
    ip_address = db.Column(db.String(20), nullable=False)
    name = db.Column(db.String(100), nullable=False)
    password = db.Column(db.String(255), nullable=False)
    comment = db.Column(db.String(255), nullable=True)
    profile = db.Column(db.String(255), nullable=True)
    profile_value = db.Column(db.String(100), nullable=True)
    is_active = db.Column(db.Boolean, default=True)
    
    # Relaciones
    router = db.relationship('Router', back_populates='secret_configs')


class ActivityLog(db.Model):
    """Modelo ActivityLog basado en tabla 'activity_logs'"""
    __tablename__ = 'activity_logs'
    
    id = db.Column(db.Integer, primary_key=True)
    router_id = db.Column(db.Integer, db.ForeignKey('routers.id'), nullable=False)
    user_id = db.Column(db.Integer, db.ForeignKey('users.id'), nullable=False)
    action = db.Column(db.String(255), nullable=False)
    description = db.Column(db.Text, nullable=True)
    ip_address = db.Column(db.String(45), nullable=True)
    user_agent = db.Column(db.Text, nullable=True)
    created_at = db.Column(db.DateTime, default=datetime.utcnow, index=True)
    
    # Relaciones
    router = db.relationship('Router', back_populates='activity_logs')
    user = db.relationship('User', back_populates='activity_logs')